From a0c1599eddff3ddfbd3d225838ac0ef07fdd3b2b Mon Sep 17 00:00:00 2001
From: mgopalakirshnan <manigandan.gopalakrishnan@lnttechservices.com>
Date: Thu, 6 Sep 2018 12:30:27 +0000
Subject: [PATCH] comcast AVEHolePuncher

Signed-off-by: mgopalakirshnan <manigandan.gopalakrishnan@lnttechservices.com>
---
 .../platform/graphics/holepunch/MediaPlayerPrivateHolePunchDummy.cpp | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/Source/WebCore/platform/graphics/holepunch/MediaPlayerPrivateHolePunchDummy.cpp b/Source/WebCore/platform/graphics/holepunch/MediaPlayerPrivateHolePunchDummy.cpp
index 76896126ba0..4b167ae9e33 100644
--- a/Source/WebCore/platform/graphics/holepunch/MediaPlayerPrivateHolePunchDummy.cpp
+++ b/Source/WebCore/platform/graphics/holepunch/MediaPlayerPrivateHolePunchDummy.cpp
@@ -46,7 +46,8 @@ static HashSet<String, ASCIICaseInsensitiveHash>& mimeTypeCache()
         return cache;
 
     const char* mimeTypes[] = {
-        "video/holepunch"
+        "video/holepunch",
+        "video/ave"
     };
 
     for (unsigned i = 0; i < (sizeof(mimeTypes) / sizeof(*mimeTypes)); ++i)
@@ -93,4 +94,4 @@ MediaPlayerPrivateHolePunchDummy::~MediaPlayerPrivateHolePunchDummy()
 
 }
 
-#endif // USE(COORDINATED_GRAPHICS_THREADED)
\ No newline at end of file
+#endif // USE(COORDINATED_GRAPHICS_THREADED)
-- 
2.16.2

