From e706429e733e097fd5eb99727f43e6f7e452eaa8 Mon Sep 17 00:00:00 2001
From: Eugene Mutavchi <Ievgen_Mutavchi@comcast.com>
Date: Tue, 2 Oct 2018 13:00:32 -0400
Subject: [PATCH] [MSE] Report 'ended' after seek

---
 .../platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp   | 1 -
 1 file changed, 1 deletion(-)

diff --git a/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp b/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp
index 8cfd74f..e3bb8a8 100644
--- a/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp
+++ b/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp
@@ -558,7 +558,6 @@ void MediaPlayerPrivateGStreamerMSE::waitForSeekCompleted()
 
 void MediaPlayerPrivateGStreamerMSE::seekCompleted()
 {
-    m_eosMarked = false;
     if (m_mseSeekCompleted)
         return;
 
-- 
2.7.4

