From 6fb42838dffe0d72813667e5bcf847e5f8ab5fae Mon Sep 17 00:00:00 2001
From: Eugene Mutavchi <Ievgen_Mutavchi@comcast.com>
Date: Fri, 26 Apr 2019 13:29:05 -0400
Subject: [PATCH] Prefer webkit decryptors

---
 .../graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp   | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp b/Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp
index 07446f9..f72abe2 100644
--- a/Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp
+++ b/Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp
@@ -483,6 +483,19 @@ GstElement* createDecryptor(const char* requestedProtectionSystemUuid)
     GstElement* decryptor = nullptr;
     GList* decryptors = gst_element_factory_list_get_elements(GST_ELEMENT_FACTORY_TYPE_DECRYPTOR, GST_RANK_MARGINAL);
 
+    // Prefer WebKit decryptors
+    decryptors = g_list_sort(decryptors, [](gconstpointer p1, gconstpointer p2) -> gint {
+        GstPluginFeature *f1, *f2;
+        const gchar* name;
+        f1 = (GstPluginFeature *) p1;
+        f2 = (GstPluginFeature *) p2;
+        if ((name = gst_plugin_feature_get_name(f1)) && g_str_has_prefix(name, "webkit"))
+            return -1;
+        if ((name = gst_plugin_feature_get_name(f2)) && g_str_has_prefix(name, "webkit"))
+            return 1;
+        return gst_plugin_feature_rank_compare_func(p1, p2);
+    });
+
     for (GList* walk = decryptors; !decryptor && walk; walk = g_list_next(walk)) {
         GstElementFactory* factory = reinterpret_cast<GstElementFactory*>(walk->data);
 
-- 
2.7.4

