From 7f07bc8ac73b5011ff0b98c1a41a1fe1f3059050 Mon Sep 17 00:00:00 2001
From: Ievgen Mutavchi <emutavchi@gmail.com>
Date: Fri, 23 Aug 2019 12:40:53 -0400
Subject: [PATCH] Detect DolbyVision

---
 .../graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp       | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp b/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp
index 71836d6..1c7f844 100644
--- a/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp
+++ b/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp
@@ -1092,6 +1092,12 @@ const static HashSet<AtomicString>& codecSet()
             set.add(AtomicString("vp9.2"));
         }
 #endif
+#if PLATFORM(BROADCOM) && USE(SVP)
+        if (gstRegistryHasElementForMediaType(videoDecoderFactories,"video/x-dvav"))
+            set.add(AtomicString("dvav*"));
+        if (gstRegistryHasElementForMediaType(videoDecoderFactories,"video/x-dvhe"))
+            set.add(AtomicString("dvhe*"));
+#endif
         gst_plugin_feature_list_free(audioDecoderFactories);
         gst_plugin_feature_list_free(videoDecoderFactories);
 
-- 
2.7.4

