From 04ad91c708bef4c012cdb2981cd9cb16f69ded85 Mon Sep 17 00:00:00 2001
From: Eugene Mutavchi <Ievgen_Mutavchi@comcast.com>
Date: Sat, 9 May 2020 16:46:21 +0000
Subject: [PATCH] Fix build of wpe test bundle
Source: COMCAST
Upstream-Status: None
Signed-off-by: Eugene Mutavchi <Ievgen_Mutavchi@comcast.com>

---
 Source/WebKit/PlatformWPE.cmake | 1 +
 1 file changed, 1 insertion(+)

diff --git a/Source/WebKit/PlatformWPE.cmake b/Source/WebKit/PlatformWPE.cmake
index 67b1b9ced47f..1ae5ea09e57e 100644
--- a/Source/WebKit/PlatformWPE.cmake
+++ b/Source/WebKit/PlatformWPE.cmake
@@ -417,6 +417,7 @@ list(APPEND WebKit_INCLUDE_DIRECTORIES
     "${WEBCORE_DIR}/dom"
     "${WEBCORE_DIR}/loader"
     "${WEBCORE_DIR}/platform"
+    "${WEBCORE_DIR}/platform/graphics"
     "${WEBCORE_DIR}/platform/graphics/cairo"
     "${WEBCORE_DIR}/platform/graphics/freetype"
     "${WEBCORE_DIR}/platform/graphics/opentype"
-- 
2.24.0

