diff --git a/CMakeLists.txt b/CMakeLists.txt
index 6f50ef5..d00c1f1 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -54,9 +54,9 @@ endif (FEATURE_SUPPORT_AKER)
 
 if (WEBCONFIG_BIN_SUPPORT)
 include_directories(${INCLUDE_DIR}/rbus
-		    ${INCLUDE_DIR}/rbus-core
 		    ${INCLUDE_DIR}/rtmessage
 )
+
 endif (WEBCONFIG_BIN_SUPPORT)
 
 if (NOT BUILD_YOCTO)
@@ -187,44 +187,22 @@ add_dependencies(liblibparodus libparodus)
 endif (FEATURE_SUPPORT_AKER)
 
 if (WEBCONFIG_BIN_SUPPORT)
-# rtMessage external dependency
-#-------------------------------------------------------------------------------
-ExternalProject_Add(rtMessage
-     PREFIX ${CMAKE_CURRENT_BINARY_DIR}/_prefix/rtMessage
-     GIT_REPOSITORY https://github.com/rdkcmf/rdk-rtmessage.git
-     GIT_TAG rdk-next
-     CMAKE_ARGS += -DBUILD_RTMESSAGE_LIB=ON
-                   -DBUILD_RTMESSAGE_SAMPLE_APP=ON
-                   -DBUILD_FOR_DESKTOP=ON
-                   -DBUILD_DATAPROVIDER_LIB=ON
-                   -DCMAKE_INSTALL_PREFIX=${INSTALL_DIR} -DBUILD_TESTING=OFF
-)
-add_library(librtMessage STATIC SHARED IMPORTED)
-add_dependencies(librtMessage rtMessage)
-
-# rbus-core external dependency	
-#-------------------------------------------------------------------------------
-ExternalProject_Add(rbus-core
-     DEPENDS rtMessage
-     PREFIX ${CMAKE_CURRENT_BINARY_DIR}/_prefix/rbus-core
-     GIT_REPOSITORY https://github.com/rdkcmf/rbuscore.git
-     GIT_TAG rdk-next
-     CMAKE_ARGS += -DBUILD_FOR_DESKTOP=ON -DCMAKE_INSTALL_PREFIX=${INSTALL_DIR}
-                   -DBUILD_TESTING=OFF
-)
-add_library(librbus-core STATIC SHARED IMPORTED)
-add_dependencies(librbus-core rbus-core)
-
 # rbus external dependency
 #-------------------------------------------------------------------------------
 ExternalProject_Add(rbus
      DEPENDS rtMessage rbus-core
      PREFIX ${CMAKE_CURRENT_BINARY_DIR}/_prefix/rbus
-     GIT_REPOSITORY https://github.com/rdkcmf/rbus.git
-     GIT_TAG rdk-next
+     GIT_REPOSITORY https://github.com/rdkcentral/rbus.git
+     GIT_TAG main
      CMAKE_ARGS += -DBUILD_FOR_DESKTOP=ON -DCMAKE_INSTALL_PREFIX=${INSTALL_DIR} -DBUILD_TESTING=OFF
 )
 
+add_library(librbuscore STATIC SHARED IMPORTED)
+add_dependencies(librbuscore rbuscore)
+
+add_library(librtMessage STATIC SHARED IMPORTED)
+add_dependencies(librtMessage rtMessage)
+
 add_library(librbus STATIC SHARED IMPORTED)
 add_dependencies(librbus rbus)
 endif (WEBCONFIG_BIN_SUPPORT)
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index ede5e29..a292659 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -62,7 +62,7 @@ set(HEADERS ${HEADERS} webcfg_rbus.h)
 set(SOURCES ${SOURCES} main.c webcfg_rbus.c)
 add_executable(webconfig ${SOURCES} ${HEADERS})
 set_property(TARGET webconfig PROPERTY C_STANDARD 99)
-set(COMMON_LIBS -lwrp-c -lpthread -lwdmp-c -lmsgpackc -ltrower-base64 -lm -lcjson -lrt -luuid -lcurl -lrbus -lrbus-core)
+set(COMMON_LIBS -lwrp-c -lpthread -lwdmp-c -lmsgpackc -ltrower-base64 -lm -lcjson -lrt -luuid -lcurl -lrbus)
 
 if (NOT DEVICE_EXTENDER)
 set(COMMON_LIBS ${COMMON_LIBS} -lcimplog)
diff --git a/src/webcfg_rbus.h b/src/webcfg_rbus.h
index 4e6df35..7203ab6 100755
--- a/src/webcfg_rbus.h
+++ b/src/webcfg_rbus.h
@@ -26,8 +26,6 @@
 #include <rbus/rbus_object.h>
 #include <rbus/rbus_property.h>
 #include <rbus/rbus_value.h>
-#include <rbus-core/rbus_core.h>
-#include <rbus-core/rbus_session_mgr.h>
 
 #include <wdmp-c.h>
 #include <wrp-c.h>
diff --git a/tests/CMakeLists.txt b/tests/CMakeLists.txt
index 4e1db1a..26bb827 100644
--- a/tests/CMakeLists.txt
+++ b/tests/CMakeLists.txt
@@ -39,7 +39,7 @@ add_executable(webcfgCli ${SOURCES})
 target_link_libraries (webcfgCli -lmsgpackc -lcurl -lpthread  -lm -luuid -ltrower-base64 -lwdmp-c -lcimplog -lcjson -lwrp-c)
 
 if (WEBCONFIG_BIN_SUPPORT)
-target_link_libraries (webcfgCli -lrbus -lrbus-core)
+target_link_libraries (webcfgCli -lrbus)
 endif (WEBCONFIG_BIN_SUPPORT)
 
 if (FEATURE_SUPPORT_AKER)
@@ -83,7 +83,7 @@ add_executable(test_multipart ${SOURCES})
 target_link_libraries (test_multipart -lcunit -lmsgpackc -lcurl -lpthread  -lm -luuid -ltrower-base64 -lwdmp-c -lcimplog -lcjson -lwrp-c)
 
 if (WEBCONFIG_BIN_SUPPORT)
-target_link_libraries (test_multipart -lrbus -lrbus-core)
+target_link_libraries (test_multipart -lrbus)
 endif (WEBCONFIG_BIN_SUPPORT)
 
 if (FEATURE_SUPPORT_AKER)
@@ -109,7 +109,7 @@ add_executable(test_mul_supp ${SOURCES})
 target_link_libraries (test_mul_supp -lcunit -lmsgpackc -lcurl -lpthread  -lm -luuid -ltrower-base64 -lwdmp-c -lcimplog -lcjson -lwrp-c)
 
 if (WEBCONFIG_BIN_SUPPORT)
-target_link_libraries (test_mul_supp -lrbus -lrbus-core)
+target_link_libraries (test_mul_supp -lrbus)
 endif (WEBCONFIG_BIN_SUPPORT)
 
 if (FEATURE_SUPPORT_AKER)
@@ -136,7 +136,7 @@ add_executable(test_events ${SOURCES})
 target_link_libraries (test_events -lcunit -lmsgpackc -lcurl -lpthread  -lm -luuid -ltrower-base64 -lwdmp-c -lcimplog -lcjson -lwrp-c)
 
 if (WEBCONFIG_BIN_SUPPORT)
-target_link_libraries (test_events -lrbus -lrbus-core)
+target_link_libraries (test_events -lrbus)
 endif (WEBCONFIG_BIN_SUPPORT)
 
 if (FEATURE_SUPPORT_AKER)
@@ -164,7 +164,7 @@ add_executable(test_events_supp  ${SOURCES})
 target_link_libraries (test_events_supp -lcunit -lmsgpackc -lcurl -lpthread  -lm -luuid -ltrower-base64 -lwdmp-c -lcimplog -lcjson -lwrp-c)
 
 if (WEBCONFIG_BIN_SUPPORT)
-target_link_libraries (test_events_supp -lrbus -lrbus-core)
+target_link_libraries (test_events_supp -lrbus)
 endif (WEBCONFIG_BIN_SUPPORT)
 
 if (FEATURE_SUPPORT_AKER)
@@ -201,7 +201,7 @@ add_executable(test_root ${SOURCES})
 target_link_libraries (test_root -lcunit -lmsgpackc -lcurl -lpthread  -lm -luuid -ltrower-base64 -lwdmp-c -lcimplog -lcjson -lwrp-c)
 
 if (WEBCONFIG_BIN_SUPPORT)
-target_link_libraries (test_root -lrbus -lrbus-core)
+target_link_libraries (test_root -lrbus)
 endif (WEBCONFIG_BIN_SUPPORT)
 
 if (FEATURE_SUPPORT_AKER)
@@ -239,7 +239,7 @@ add_executable(test_multipart_unittest ${SOURCES})
 target_link_libraries (test_multipart_unittest -lcunit -lmsgpackc -lcurl -lpthread  -lm -luuid -ltrower-base64 -lwdmp-c -lcimplog -lcjson -lwrp-c)
 
 if (WEBCONFIG_BIN_SUPPORT)
-target_link_libraries (test_multipart_unittest -lrbus -lrbus-core)
+target_link_libraries (test_multipart_unittest -lrbus)
 endif (WEBCONFIG_BIN_SUPPORT)
 
 if (FEATURE_SUPPORT_AKER)
@@ -254,7 +254,7 @@ target_link_libraries (test_multipart_unittest gcov -Wl,--no-as-needed )
 if (WEBCONFIG_BIN_SUPPORT)
 add_test(NAME test_rbus_fr COMMAND ${MEMORY_CHECK} ./test_rbus_fr)
 add_executable(test_rbus_fr test_rbus_fr.c ../src/webcfg_rbus.c )
-target_link_libraries (test_rbus_fr -lcunit -lwrp-c -lcimplog -lmsgpackc -lcurl -lpthread  -lm -luuid -ltrower-base64 -lwdmp-c -lcjson  -lrbus -lrbus-core)
+target_link_libraries (test_rbus_fr -lcunit -lwrp-c -lcimplog -lmsgpackc -lcurl -lpthread  -lm -luuid -ltrower-base64 -lwdmp-c -lcjson  -lrbus)
 
 target_link_libraries (test_rbus_fr gcov -Wl,--no-as-needed )
 endif (WEBCONFIG_BIN_SUPPORT)
