From e76fd45a4e9cc2c89c3caa1aeb4663e4ed1ab767 Mon Sep 17 00:00:00 2001
From: Jaga <jagadheesan_duraisamy@comcast.com>
Date: Thu, 28 Jan 2021 13:19:21 +0000
Subject: [PATCH] glibc 2.31 ignore truncated dns response

---
 resolv/resolv.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/resolv/resolv.h b/resolv/resolv.h
index a039a9e636..70437fd263 100644
--- a/resolv/resolv.h
+++ b/resolv/resolv.h
@@ -133,7 +133,7 @@ struct res_sym {
 #define RES_NORELOAD    0x02000000 /* No automatic configuration reload.  */
 #define RES_TRUSTAD     0x04000000 /* Request AD bit, keep it in responses.  */
 
-#define RES_DEFAULT	(RES_RECURSE|RES_DEFNAMES|RES_DNSRCH)
+#define RES_DEFAULT	(RES_IGNTC|RES_RECURSE|RES_DEFNAMES|RES_DNSRCH)
 
 /*
  * Resolver "pfcode" values.  Used by dig.
