Date: Aug 17, 2017 6:48 AM

From: bb18136252c2d779c0a96a7ac582ea917cdac1be Aug 22, 2017 11:01 AM

Subject: [PATCH] RDKB-13471: Integration of smcroute

	Integration of smcroute for multicast routing

Source: COMCAST


Upstream-Status: Pending

Signed-off-by: shirish <shirish_s@infosys.com>
---

diff --git a/mroute-api.c b/mroute-api.c
index f283de5..0228bcb 100644
--- a/mroute-api.c
+++ b/mroute-api.c
@@ -547,8 +547,11 @@ static int mroute6_add_mif(struct iface *iface)
 	       iface->name, mc.mif6c_mifi, mc.mif6c_pifi, mc.mif6c_flags);
 
 	if (setsockopt(mroute6_socket, IPPROTO_IPV6, MRT6_ADD_MIF, (void *)&mc, sizeof(mc)))
+	{
+		smclog(LOG_WARNING, errno, "Failed adding MIF for iface %s", iface->name);
+	if (setsockopt(mroute6_socket, IPPROTO_IPV6, SO_REUSEADDR, (void *)&mc, sizeof(mc)))
 		smclog(LOG_ERR, errno, "Failed adding MIF for iface %s", iface->name);
-
+	}
 	iface->mif = mif;
 	mif_list[mif].iface = iface;
 
