Index: dropbear-2017.75/dbutil.c
===================================================================
--- dropbear-2017.75.orig/dbutil.c
+++ dropbear-2017.75/dbutil.c
@@ -68,6 +68,7 @@
 #include "atomicio.h"
 
 #define MAX_FMT 100
+#define SSH_SHELL_LOG "/opt/secure/rdk_shell.log"
 
 static void generic_dropbear_exit(int exitcode, const char* format, 
 		va_list param) ATTRIB_NORETURN;
@@ -148,6 +149,31 @@ void dropbear_log(int priority, const ch
 	va_end(param);
 }
 
+void ssh_logging(const char* format, ...) {
+#ifdef RDK_BROADBAND
+    if ((access("/rdklogs/logs/rdk_shell.log", F_OK ) != 0)) {
+        symlink("/opt/secure/rdk_shell.log","/rdklogs/logs/rdk_shell.log");
+    }
+#else
+    if ((access("/opt/logs/rdk_shell.log", F_OK ) != 0)) {
+        symlink("/opt/secure/rdk_shell.log","/opt/logs/rdk_shell.log");
+    }
+#endif
+    FILE* fp = NULL;
+    fp = fopen(SSH_SHELL_LOG, "a+");
+    if (fp != NULL) {
+        char printbuf[1024];
+        va_list param;
+        va_start(param, format);
+        vsnprintf(printbuf, sizeof(printbuf), format, param);
+        fprintf(fp, "%s", printbuf);
+        va_end(param);
+        fclose(fp);
+    }
+    else {
+        fprintf(stderr, "Error opening file \n");
+    }
+}
 
 #ifdef DEBUG_TRACE
 
Index: dropbear-2017.75/svr-chansession.c
===================================================================
--- dropbear-2017.75.orig/svr-chansession.c
+++ dropbear-2017.75/svr-chansession.c
@@ -674,6 +674,11 @@ static int sessioncommand(struct Channel
 	/* take public key option 'command' into account */
 	svr_pubkey_set_forced_command(chansess);
 
+        /* To record the commands executed from ssh */
+        if (chansess->cmd) {
+                ssh_logging("%s\n", chansess->cmd);
+        }
+
 #ifdef LOG_COMMANDS
 	if (chansess->cmd) {
 		dropbear_log(LOG_INFO, "User %s executing '%s'", 
