Index: dropbear-2017.75/svr-authpubkey.c
===================================================================
--- dropbear-2017.75.orig/svr-authpubkey.c
+++ dropbear-2017.75/svr-authpubkey.c
@@ -64,6 +64,7 @@
 #include "ssh.h"
 #include "packet.h"
 #include "algo.h"
+#include <telemetry_busmessage_sender.h>
 
 #ifdef ENABLE_SVR_PUBKEY_AUTH
 
@@ -143,11 +144,13 @@ void svr_auth_pubkey() {
 	/* ... and finally verify the signature */
 	fp = sign_key_fingerprint(keyblob, keybloblen);
 	if (buf_verify(ses.payload, key, signbuf) == DROPBEAR_SUCCESS) {
+                t2_event_s("SSH_connection_split", ses.authstate.pw_name);
 		dropbear_log(LOG_NOTICE,
 				"Pubkey auth succeeded for '%s' with key %s from %s",
 				ses.authstate.pw_name, fp, svr_ses.addrstring);
 		send_msg_userauth_success();
 	} else {
+                t2_event_d("Illegal_SSH_connection",1);
 		dropbear_log(LOG_WARNING,
 				"Pubkey auth bad signature for '%s' with key %s from %s",
 				ses.authstate.pw_name, fp, svr_ses.addrstring);
