From 6f4065fb7701085d417f15f0d962f8ba82f20d99 Mon Sep 17 00:00:00 2001
From: Eugene Mutavchi <Ievgen_Mutavchi@comcast.com>
Date: Thu, 31 Mar 2022 12:28:24 +0000
Subject: [PATCH] Disable directcall optimisation for MIPS

---
 Source/JavaScriptCore/dfg/DFGStrengthReductionPhase.cpp | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/Source/JavaScriptCore/dfg/DFGStrengthReductionPhase.cpp b/Source/JavaScriptCore/dfg/DFGStrengthReductionPhase.cpp
index 1313b91514a9..03b069901006 100644
--- a/Source/JavaScriptCore/dfg/DFGStrengthReductionPhase.cpp
+++ b/Source/JavaScriptCore/dfg/DFGStrengthReductionPhase.cpp
@@ -896,6 +896,9 @@ private:
         case Construct:
         case TailCallInlinedCaller:
         case TailCall: {
+            #if CPU(MIPS)
+            break;
+            #endif
             ExecutableBase* executable = nullptr;
             Edge callee = m_graph.varArgChild(m_node, 0);
             CallVariant callVariant;
-- 
2.25.1

