From 454edc676a2793dd212f46f71d010bac7c648e1c Mon Sep 17 00:00:00 2001
From: kkanag314 <krishnapriya_kanagaraj@comcast.com>
Date: Wed, 14 Jul 2021 08:54:59 +0000
Subject: [PATCH] RWI works for one app session then fail

---
 .../UIProcess/InspectorServer/WebInspectorServer.cpp     | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/Source/WebKit/UIProcess/InspectorServer/WebInspectorServer.cpp b/Source/WebKit/UIProcess/InspectorServer/WebInspectorServer.cpp
index 12295f30c8aa..691f0478a87d 100644
--- a/Source/WebKit/UIProcess/InspectorServer/WebInspectorServer.cpp
+++ b/Source/WebKit/UIProcess/InspectorServer/WebInspectorServer.cpp
@@ -152,10 +152,15 @@ bool WebInspectorServer::shouldInhibitLocalHostInspection(int pageId)
     WebInspectorProxy* client = m_clientMap.get(pageId);
     if(client && client->inspectedPage())
     {
-        if(client->inspectedPage()->pageLoadState().activeURL().find("localhost:") != notFound)
+        if(client->inspectedPage()->pageLoadState().activeURL().find("localhost:") != notFound) {
+            m_clientMap.remove(pageId);
             return true;
-        if(client->inspectedPage()->pageLoadState().activeURL().find("file://") != notFound)
+        }
+
+        if(client->inspectedPage()->pageLoadState().activeURL().find("file://") != notFound) {
+            m_clientMap.remove(pageId);
             return true;
+        }
     }
     else
         return true;
-- 
2.29.0

