--- wpe-webkit/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp
+++ wpe-webkit/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp
@@ -2999,14 +3004,17 @@
 #if PLATFORM(BROADCOM)
     if (g_str_has_prefix(GST_ELEMENT_NAME(element), "brcmaudiosink")) {
         g_object_set(G_OBJECT(element), "async", TRUE, nullptr);
     }
     else if (g_str_has_prefix(GST_ELEMENT_NAME(element), "brcmaudiodecoder")) {
         g_object_set(G_OBJECT(element), "audio_pts_disco_threshold", G_MAXUINT, nullptr);
-    }
-
+        if (player->isLiveStream() && !player->m_hasVideo) {
+            /* if iHeart, reduce audio buffering to 500ms, impacts start up */
+            g_object_set(G_OBJECT(element), "limit_buffering_ms", 500, "limit_buffering", 1, NULL);
+        }
+    }
 #endif
 
 #if USE(WESTEROS_SINK)
     static GstCaps *westerosSinkCaps = nullptr;
     static GType westerosSinkType = G_TYPE_INVALID;
     static std::once_flag onceFlag;
