Date: Wed Dec 30 18:06:10 2020 +0000

From: I540fd7153170c6d8cae313eaf0cbabf8acd43644 Wed Dec 30 18:06:10

Subject: DELIA-47925 : syssnmpagent crash with signature libc.so.6 reason SIGABRT

Source: COMCAST

Upstream-Status:N/A

Signed-off-by: Rense Jacob <Rense_Jacob2@comcast.com>

Index: net-snmp-5.8/agent/agent_registry.c
===================================================================
--- net-snmp-5.8.orig/agent/agent_registry.c
+++ net-snmp-5.8/agent/agent_registry.c
@@ -800,12 +800,11 @@ netsnmp_subtree_load(netsnmp_subtree *ne
     /*  Handle new subtrees that start in virgin territory.  */
 
     if (tree1 == NULL) {
-        /*netsnmp_subtree *new2 = NULL;*/
+        netsnmp_subtree *new2 = NULL;
 	/*  Is there any overlap with later subtrees?  */
 	if (tree2 && snmp_oid_compare(new_sub->end_a, new_sub->end_len,
 				      tree2->start_a, tree2->start_len) > 0) {
-	    /*new2 =*/
-            netsnmp_subtree_split(new_sub, tree2->start_a, tree2->start_len);
+	    new2 = netsnmp_subtree_split(new_sub, tree2->start_a, tree2->start_len);
 	}
 
 	/*  Link the new subtree (less any overlapping region) with the list of
@@ -827,7 +826,6 @@ netsnmp_subtree_load(netsnmp_subtree *ne
 
             netsnmp_subtree_change_next(new_sub, tree2);
 
-#if 0
             /* The code below cannot be reached which is why it has been
                surrounded with #if 0 / #endif. */
 	    /* If there was any overlap, recurse to merge in the overlapping
@@ -835,7 +833,6 @@ netsnmp_subtree_load(netsnmp_subtree *ne
 	    if (new2) {
 		return netsnmp_subtree_load(new2, context_name);
 	    }
-#endif
 	}
     } else {
 	/*  If the new subtree starts *within* an existing registration

