Date: Nov 27 10:50:30 2017

From: 0cadb71747b08193f24d2c501b66488024bfc071 Jan 15 16:38:29 2018

Subject: [PATCH] RDKB-14741: Remove lighttpd hook from meta-rdk

	Remove lighttpd hook from meta-rdk.

Source: COMCAST

Upstream-Status: Pending

Signed-off-by: Deva Thiyagarajan <Deva_Thiyagarajan@comcast.com>
---

Index: lxc-1.1.2/hooks/Makefile.am
===================================================================
--- lxc-1.1.2.orig/hooks/Makefile.am
+++ lxc-1.1.2/hooks/Makefile.am
@@ -5,6 +5,7 @@ hooks_SCRIPTS = \
 	mountecryptfsroot \
 	ubuntu-cloud-prep \
 	squid-deb-proxy-client \
-	rdksystemdhook
+	rdksystemdhook \
+	lighttpd_systemdhook
 
 EXTRA_DIST=$(hooks_SCRIPTS)
