From 05988387c59e71520b62b50eb624c2f9386ef461 Mon Sep 17 00:00:00 2001
From: Eugene Mutavchi <Ievgen_Mutavchi@comcast.com>
Date: Thu, 23 Sep 2021 00:00:28 +0000
Subject: [PATCH] [GST] return cached position when sink returns invalid one

---
 .../graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp      | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp b/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp
index f697fb63bc5c..1d223029c3a8 100644
--- a/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp
+++ b/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp
@@ -363,6 +363,8 @@ MediaTime MediaPlayerPrivateGStreamer::playbackPosition() const
             return m_seekTime;
 
         MediaTime duration = durationMediaTime();
+        if (m_cachedPosition.isValid() && duration.isValid() && m_cachedPosition > duration)
+            return m_cachedPosition;
         return duration.isInvalid() ? MediaTime::zeroTime() : duration;
     }
 
@@ -403,6 +405,8 @@ MediaTime MediaPlayerPrivateGStreamer::playbackPosition() const
         playbackPosition = MediaTime(gstreamerPosition, GST_SECOND);
     else if (m_canFallBackToLastFinishedSeekPosition)
         playbackPosition = m_seekTime;
+    else if (m_cachedPosition.isValid())
+        playbackPosition = m_cachedPosition;
 
     m_playbackProgress = m_cachedPosition.isValid() ? abs(playbackPosition - m_cachedPosition) : playbackPosition;
     m_cachedPosition = playbackPosition;
@@ -2300,7 +2304,7 @@ void MediaPlayerPrivateGStreamer::didEnd()
     // Synchronize position and duration values to not confuse the
     // HTMLMediaElement. In some cases like reverse playback the
     // position is not always reported as 0 for instance.
-    m_cachedPosition = MediaTime::invalidTime();
+    m_lastQueryTime.reset();
     MediaTime now = currentMediaTime();
     if (now > MediaTime { } && now <= durationMediaTime())
         m_player->durationChanged();
-- 
2.25.1

