From 7addea0c663356d9b3c5e948761f0307385394e8 Mon Sep 17 00:00:00 2001
From: nganes961 <nambirajan_ganesan@comcast.com>
Date: Fri, 15 Oct 2021 16:34:32 +0000
Subject: [PATCH] Temporary disable cookies in gst context

Signed-off-by: nganes961 <nambirajan_ganesan@comcast.com>
---
 .../platform/graphics/gstreamer/WebKitWebSourceGStreamer.cpp    | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Source/WebCore/platform/graphics/gstreamer/WebKitWebSourceGStreamer.cpp b/Source/WebCore/platform/graphics/gstreamer/WebKitWebSourceGStreamer.cpp
index 9820b5735b2f..12f15e4fcff8 100644
--- a/Source/WebCore/platform/graphics/gstreamer/WebKitWebSourceGStreamer.cpp
+++ b/Source/WebCore/platform/graphics/gstreamer/WebKitWebSourceGStreamer.cpp
@@ -536,7 +536,7 @@ static void webKitWebSrcStart(WebKitWebSrc* src)
             priv->loader = priv->player->createResourceLoader();
 
         {
-            String cookies = WebCore::cookies(*priv->player->cachedResourceLoader()->document(), request.url());
+            String cookies; // = WebCore::cookies(*priv->player->cachedResourceLoader()->document(), request.url());
             priv->cookies = GUniquePtr<gchar>(g_strdup(cookies.utf8().data()));
         }
 
-- 
2.17.1

