--- gstreamer/libs/gst/check/gstcheck.c	2021-11-24 06:15:05.008600414 +0000
+++ gstreamer_modified/libs/gst/check/gstcheck.c	2021-11-24 06:13:49.848389676 +0000
@@ -347,6 +347,7 @@
   guint timeout_multiplier = 1;
   GOptionContext *ctx;
   GError *err = NULL;
+  GLogLevelFlags _gst_log_level = G_LOG_LEVEL_CRITICAL | G_LOG_LEVEL_WARNING;
   GOptionEntry options[] = {
     {"list-tests", 'l', 0, G_OPTION_ARG_NONE, &_gst_check_list_tests,
         "List tests present in the testsuite", NULL},
@@ -375,18 +376,16 @@
   if (g_getenv ("GST_TEST_DEBUG"))
     _gst_check_debug = TRUE;
 
+  if (g_getenv ("GST_LOG_LEVEL"))
+    _gst_log_level = atoi(g_getenv ("GST_LOG_LEVEL"));
+
   g_log_set_handler (NULL, G_LOG_LEVEL_MESSAGE, gst_check_log_message_func,
       NULL);
-  g_log_set_handler (NULL, G_LOG_LEVEL_CRITICAL | G_LOG_LEVEL_WARNING,
-      gst_check_log_critical_func, NULL);
-  g_log_set_handler ("GStreamer", G_LOG_LEVEL_CRITICAL | G_LOG_LEVEL_WARNING,
-      gst_check_log_critical_func, NULL);
-  g_log_set_handler ("GLib-GObject", G_LOG_LEVEL_CRITICAL | G_LOG_LEVEL_WARNING,
-      gst_check_log_critical_func, NULL);
-  g_log_set_handler ("GLib-GIO", G_LOG_LEVEL_CRITICAL | G_LOG_LEVEL_WARNING,
-      gst_check_log_critical_func, NULL);
-  g_log_set_handler ("GLib", G_LOG_LEVEL_CRITICAL | G_LOG_LEVEL_WARNING,
-      gst_check_log_critical_func, NULL);
+  g_log_set_handler (NULL, _gst_log_level, gst_check_log_critical_func, NULL);
+  g_log_set_handler ("GStreamer", _gst_log_level, gst_check_log_critical_func, NULL);
+  g_log_set_handler ("GLib-GObject", _gst_log_level, gst_check_log_critical_func, NULL);
+  g_log_set_handler ("GLib-GIO", _gst_log_level, gst_check_log_critical_func, NULL);
+  g_log_set_handler ("GLib", _gst_log_level, gst_check_log_critical_func, NULL);
   g_test_log_set_fatal_handler (gst_check_log_fatal_func, NULL);
 
   print_plugins ();
