Date: Aug 23 14:44:26 2016

From: eace82070f519ed86a2869dbdfe437f407ef282d Aug 24 11:04:58 2016 

Subject: [PATCH] RDK-16104 : Develop code - Brotli

	Enable brotli and solve compile time issues
	
Source: COMCAST

Upstream-Status: Pending

Signed-off-by: Vargheese Baby <vargheesebaby@tataelxsi.com>
---

Index: git/MANIFEST.in
===================================================================
--- git.orig/MANIFEST.in
+++ git/MANIFEST.in
@@ -7,9 +7,6 @@ include enc/*.c
 include enc/*.h
 include LICENSE
 include MANIFEST.in
-include python/bro.py
-include python/brotlimodule.cc
-include python/README.md
 include README.md
 include setup.py
 include tools/bro.c
