Date: Mar 18, 2020 11:43 AM
From: 1ad675b8532f329bebaf40cda54192d0cb4feecb Feb 26, 2020 10:31 PM
Subject: IPv6 is not working, so use IPv4
Source: COMCAST
Upstream-Status: Pending
Signed-off-by: Thomas Lea <thomas_lea@comcast.com>
---
diff --git a/ports/common/src/iot_network_openssl.c b/ports/common/src/iot_network_openssl.c
index a57fa0b..174e605 100644
--- a/ports/common/src/iot_network_openssl.c
+++ b/ports/common/src/iot_network_openssl.c
@@ -218,10 +218,13 @@ static int _dnsLookup( IotNetworkServerInfo_t pServerInfo )
     struct addrinfo * pListHead = NULL, * pAddressInfo = NULL;
     struct sockaddr * pServer = NULL;
     socklen_t serverLength = 0;
+    struct addrinfo hints;
 
     /* Perform a DNS lookup of host name. */
     IotLogDebug( "Performing DNS lookup of %s", pServerInfo->pHostName );
-    status = getaddrinfo( pServerInfo->pHostName, NULL, NULL, &pListHead );
+    memset(&hints, 0, sizeof(struct addrinfo));
+    hints.ai_family = AF_INET;
+    status = getaddrinfo( pServerInfo->pHostName, NULL, &hints, &pListHead );
 
     if( status != 0 )
     {
