From 186c0a6dc0e59d7e8127c24499219dcc1ad16a7b Mon Sep 17 00:00:00 2001
From: Eugene Mutavchi <Ievgen_Mutavchi@comcast.com>
Date: Tue, 11 Jan 2022 21:07:17 +0000
Subject: [PATCH] comcast - WKIT-553 - add video/ave mimetype for holepunch

Signed-off-by: Eugene Mutavchi <Ievgen_Mutavchi@comcast.com>
---
 .../graphics/holepunch/MediaPlayerPrivateHolePunch.cpp         | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/Source/WebCore/platform/graphics/holepunch/MediaPlayerPrivateHolePunch.cpp b/Source/WebCore/platform/graphics/holepunch/MediaPlayerPrivateHolePunch.cpp
index 56c8b92d96dc..4f551e68b86b 100644
--- a/Source/WebCore/platform/graphics/holepunch/MediaPlayerPrivateHolePunch.cpp
+++ b/Source/WebCore/platform/graphics/holepunch/MediaPlayerPrivateHolePunch.cpp
@@ -108,7 +108,8 @@ static HashSet<String, ASCIICaseInsensitiveHash>& mimeTypeCache()
         return cache;
 
     const char* mimeTypes[] = {
-        "video/holepunch"
+        "video/holepunch",
+        "video/ave"
     };
 
     for (unsigned i = 0; i < (sizeof(mimeTypes) / sizeof(*mimeTypes)); ++i)
-- 
2.25.1

