##################################################################
Date: Tue Feb 18 17:12:03 2020 +0000
From: 
Subject: [PATCH] Create storage directory before starting service 
Source: SKY
Upstream-Status: Pending
Signed-off-by: Tijo Thomas John <tijo.thomas@sky.uk>
##################################################################

Index: bluez-5.48/src/bluetooth.service.in
===================================================================
--- bluez-5.48.orig/src/bluetooth.service.in
+++ bluez-5.48/src/bluetooth.service.in
@@ -8,6 +8,8 @@ ConditionPathIsDirectory=/sys/class/blue
 [Service]
 Type=dbus
 BusName=org.bluez
+ExecStartPre=/bin/mkdir -p /var/lib/bluetooth
+ExecStartPre=/bin/mkdir -p /opt/lib/bluetooth
 ExecStart=@libexecdir@/bluetoothd
 NotifyAccess=main
 #WatchdogSec=10
